File: front_igm/CustomerView/models/Order.js
import { Line } from "./Line.js";

export class Order{
    static VALID_STATUS = ["pending","paid","canceled","expired"];

    constructor(id,client_email,status,lines,currency){
        this.id = id; // str
        this.client_email = client_email; // str
        this.status = Order.VALID_STATUS[status]; // int
        this.lines = lines ?? []; // list<Lines> 
        this.currency = currency; // str
    }

    getTotal(){
        if (!this.lines || this.lines.length === 0) return 0;

        return this.lines.reduce((acc, line) => {
            return acc + (line.unit_price * line.quantity);
        }, 0);
    }

    addLine(line){
        if (!line) return;

        if (!this.lines) this.lines = [];

        const productId = line.product?.id;
        if (productId === undefined || productId === null) return;

        const existingLine = this.lines.find(l => l.product?.id === productId);

        if (existingLine){
            existingLine.quantity += line.quantity;
            return;
        }

        this.lines.push(line);
    }

    removeLineByProductId(productId){
        if (!this.lines || this.lines.length === 0) return;

        this.lines = this.lines.filter(line => line.product?.id !== productId);
    }

    removeAllLines(){
        this.lines = [];
    }

    save(){};
    static getAllOrders(){};
    static getOrderByEmail(){};

    // json handlers
    static fromJson(json){
            if (!json) return null;

            const lines = (json.lines ?? []).map(l => Line.fromJson(l));

            // status se guarda como string, pero tu constructor espera index
            const statusIndex = Order.VALID_STATUS.indexOf(json.status);

            return new Order(
                json.id,
                json.client_email,
                statusIndex >= 0 ? statusIndex : 0,
                lines,
                json.currency
            );
    };
    toJson(){
        return {
            "id": this.id,
            "client_email": this.client_email,
            "status": this.status,
            "lines": (this.lines ?? []).map(line => line.toJson()),
            "currency": this.currency
        };
    };

}

File: back_igm/core/models/order.py
from typing import List, Dict, Any
from line import Line


class Order:
    VALID_STATUS = ["pending", "paid", "canceled", "expired"]
    CURRENCYS = {"ARS", "USD"}

    def __init__(self, id: str, client_email: str, status: int, lines: List[Line], currency: str):
        self.id = id
        self.client_email = client_email
        self.status = Order.VALID_STATUS[status]
        self.lines = lines
        self.currency = currency

    @staticmethod
    def fromJson(data: Dict[str, Any]) -> "Order":
        return Order(
            id=str(data["id"]),
            client_email=str(data["client_email"]),
            status=int(data["status"]),
            lines=[Line.fromJson(line) for line in data["lines"]],
            currency=str(data["currency"])
        )

    def toJson(self) -> Dict[str, Any]:
        return {
            "id": self.id,
            "client_email": self.client_email,
            "status": self.status,
            "lines": [line.toJson() for line in self.lines],
            "currency": self.currency
        }
    
    def getTotal(self) -> float:
        if not self.lines:
            return 0.0

        return sum(line.product.price * line.quantity for line in self.lines)

    def addLine(self, line: Line) -> None:
        if not line:
            return

        if not self.lines:
            self.lines = []

        product_id = getattr(line.product, "id", None)
        if product_id is None:
            return

        for existing_line in self.lines:
            if getattr(existing_line.product, "id", None) == product_id:
                existing_line.quantity += line.quantity
                return

        self.lines.append(line)

    def removeLineByProductId(self, product_id: str) -> None:
        if not self.lines:
            return

        self.lines = [
            line for line in self.lines
            if getattr(line.product, "id", None) != product_id
        ]

    def removeAllLines(self) -> None:
        self.lines = []